models/user.go

```diff
@@ -0,0 +1,54 @@
+package models
+
+import (
+	"time"
+
+	"github.com/astaxie/beego/orm"
+	"github.com/astaxie/beego/validation"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type User struct {
+	Id             int
+	Email          string `orm:"unique"`
+	HashedPassword string
+
+	CreatedAt time.Time `orm:"auto_now_add;type(datetime)"`
+	UpdatedAt time.Time `orm:"auto_now;type(datetime)"`
+}
+
+func init() {
+	orm.RegisterModel(new(User))
+}
+
+func (u *User) TableName() string {
+	return "users"
+}
+
+func CreateUser(user *User, password string) error {
+	o := orm.NewOrm()
+
+	hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	user.HashedPassword = string([]byte(hashedPassword))
+
+	valid := validate(user, password)
+
+	if valid.HasErrors() {
+		return valid.Errors[0]
+	} else {
+		_, err := o.Insert(user)
+
+		return err
+	}
+}
+
+func validate(user *User, password string) validation.Validation {
```

Comment:
Could you please explain why you decide to execute this function in the model? IMO this is the user input validation and should be executed in the controller.
cc @olivierobert

Reply:
This is the data validation, like validation before adding the data into DB.

The data could be added into the database not only from Controller, so that why we should add the validation step at the model level.

Does that make sense?

Reply:
Validations can live in both models and services (and controllers some time). But the validations have different purposes. For models, it's about data integrity at the database level. In services/use cases/form objects, it's about validating the user input in a specific case. The best is to have validation in both places and model validation to be lightweight (based on the DB column formats) leaving complex validations to service/use cases/form objects.

Reply:
Thanks for clarify @andyduong1920 @olivierobert